--- "farmhash.orig/src\\farmhash.cc"	2019-01-26 18:21:59.097602200 +0800
+++ "farmhash/src\\farmhash.cc"	2019-01-29 21:01:34.119602200 +0800
@@ -114,7 +114,7 @@
 
 #if defined(FARMHASH_UNKNOWN_ENDIAN) || !defined(bswap_64)
 
-#ifdef _MSC_VER
+#ifdef _WIN32
 
 #undef bswap_32
 #undef bswap_64
@@ -216,7 +216,7 @@
   return shift == 0 ? val : ((val >> shift) | (val << (64 - shift)));
 }
 
-#if defined(_MSC_VER) && defined(FARMHASH_ROTR)
+#if defined(_WIN32) && defined(FARMHASH_ROTR)
 
 STATIC_INLINE uint32_t Rotate32(uint32_t val, int shift) {
   return sizeof(unsigned long) == sizeof(val) ?
