--- png.orig/CMakeLists.txt	2019-01-26 18:20:01.178602200 +0800
+++ png/CMakeLists.txt	2019-01-26 18:20:09.905602200 +0800
@@ -776,17 +776,17 @@
     # If we don't need to symlink something that's coming from a build target,
     # we can go ahead and symlink/copy at configure time.
 
-    if(CMAKE_HOST_WIN32 AND NOT CYGWIN AND NOT MSYS)
+    if(CMAKE_HOST_WIN32 AND NOT CYGWIN AND (NOT MSYS OR MINGW))
       execute_process(
 	COMMAND "${CMAKE_COMMAND}" -E copy_if_different ${S_FILE} ${DEST_FILE}
 	WORKING_DIRECTORY "${CMAKE_CURRENT_BINARY_DIR}"
 	)
-    else(CMAKE_HOST_WIN32 AND NOT CYGWIN AND NOT MSYS)
+    else(CMAKE_HOST_WIN32 AND NOT CYGWIN AND (NOT MSYS OR MINGW))
       execute_process(
 	COMMAND ${CMAKE_COMMAND} -E create_symlink ${S_FILE} ${DEST_FILE}
 	WORKING_DIRECTORY "${CMAKE_CURRENT_BINARY_DIR}"
 	)
-    endif(CMAKE_HOST_WIN32 AND NOT CYGWIN AND NOT MSYS)
+    endif(CMAKE_HOST_WIN32 AND NOT CYGWIN AND (NOT MSYS OR MINGW))
   endif(S_FILE)
 
   if(S_TARGET)
@@ -794,15 +794,15 @@
     # simplicity make the symlink a POST_BUILD step and use the TARGET
     # signature of add_custom_command.
 
-    if(CMAKE_HOST_WIN32 AND NOT CYGWIN AND NOT MSYS)
+    if(CMAKE_HOST_WIN32 AND NOT CYGWIN AND (NOT MSYS OR MINGW))
       add_custom_command(TARGET ${S_TARGET} POST_BUILD
 	COMMAND "${CMAKE_COMMAND}" -E copy_if_different $<TARGET_LINKER_FILE_NAME:${S_TARGET}> $<TARGET_LINKER_FILE_DIR:${S_TARGET}>/${DEST_FILE}
 	)
-    else(CMAKE_HOST_WIN32 AND NOT CYGWIN AND NOT MSYS)
+    else(CMAKE_HOST_WIN32 AND NOT CYGWIN AND (NOT MSYS OR MINGW))
       add_custom_command(TARGET ${S_TARGET} POST_BUILD
 	COMMAND "${CMAKE_COMMAND}" -E create_symlink $<TARGET_LINKER_FILE_NAME:${S_TARGET}> $<TARGET_LINKER_FILE_DIR:${S_TARGET}>/${DEST_FILE}
 	)
-    endif(CMAKE_HOST_WIN32 AND NOT CYGWIN AND NOT MSYS)
+    endif(CMAKE_HOST_WIN32 AND NOT CYGWIN AND (NOT MSYS OR MINGW))
 
   endif(S_TARGET)
 
